File: Leetcode_JS/longest-substring-without-repeat-char.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    // Info
    /*
        - We need to find a substring in string s
        - the substring needs to contain different character (NO DUPLICATES)
        - return longest sub
     */

    //  Plan
    /*
        - Use pointers to keep track of different characters in a row
        - use maxLen to store result
        - Iterate over the string
            - Update maxLen with the next longest contiguous sub
        - return maxLen
     */

     let pointer1 = 0
     let newSet = new Set()
     let maxLen = 0
     
     if (s.length === 0) return maxLen

     for (let pointer2 = 0; pointer2 < s.length; pointer2++) {
        while (newSet.has(s[pointer2])) {
            newSet.delete(s[pointer1])
            pointer1++
        }
       
       newSet.add(s[pointer2])
       maxLen = Math.max(maxLen, pointer2 - pointer1 + 1)
     }

     return maxLen
};

File: Leetcode_Python/longest-substring-without-repeat-char.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        new_set = set()
        pointer1 = 0
        max_len = 0

        for pointer2 in range(len(s)):
            while s[pointer2] in new_set:
                new_set.remove(s[pointer1])
                pointer1+=1

            new_set.add(s[pointer2])
            max_len = max(max_len, pointer2 - pointer1 + 1)

        return max_len